### PR TITLE
Ml pipeline update

### LINKED ISSUE #43: Demonstrate use of hyperdrive for parallelized hyperparameter tuning

### LINKED ISSUE #46: Update pipeline to use new sample
Where/how to save test dataset to be stored seperately?

### PR DESCRIPTION
**Pull request by hannahbrown7**
This pull request covers: 
- incorporating new sampling functions into prd_pipeline.py to save out a test dataset (need to optimise where this is saved to!), then return train and validate datasets.
- demonstrate the use of hyperdrive for distributed hyper parameter tuning, including removing hard coded hyper parameters in prd_cluster_train_demo.py

I have made minor changes to the prd_mlflow_on_azure_demo and prd_mlops_azml_cluster_demo notebooks which make use of prd_pipeline so that they still work with the updated sampling functions

closes #43 
closes #46

### CODE REVIEW

**Review by stevehadd — COMMENTED** [review]
Here they are!

File: model_pipeline/prd_mlflow_on_azure_demo.ipynb [hunk 1]
```diff
@@ -65,23 +65,15 @@
    "execution_count": 5,
```
Comment (stevehadd) [inline model_pipeline/prd_mlflow_on_azure_demo.ipynb:65]: <p>If this cell is no longer used, I think we can happily delete it from the notebook.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/informatics-lab/precip_rediagnosis/pull/50/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

File: model_pipeline/prd_mlops_azml_cluster_hyperdrive_demo.ipynb [hunk 2]
```diff
@@ -0,0 +1,560 @@
+{
```
Comment (stevehadd) [inline model_pipeline/prd_mlops_azml_cluster_hyperdrive_demo.ipynb:1]: <p>I think its generally good practice to commit the Jupyter Notebook to Github with the cells having been run, so that you can see output with the code and other text, which is a key advantage of notebooks over code.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/informatics-lab/precip_rediagnosis/pull/50/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (stevehadd) [inline model_pipeline/prd_mlops_azml_cluster_hyperdrive_demo.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #8.        }
</pre><p>Why did you choose random sampling? This may just have been been the easiest option to start with.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/informatics-lab/precip_rediagnosis/pull/50/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (stevehadd) [inline model_pipeline/prd_mlops_azml_cluster_hyperdrive_demo.ipynb:1]: <p>I was thinking we should maybe try to store these values in a config file somewhere, maybe in a folder in the repo? What do you think, would that be too much fuss?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/informatics-lab/precip_rediagnosis/pull/50/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (stevehadd) [inline model_pipeline/prd_mlops_azml_cluster_hyperdrive_demo.ipynb:1]: <p>It would be good to add one or more cells which explain the hyperparameters used to configure the hyperparameter tuning, and general how this way of HPT works.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/informatics-lab/precip_rediagnosis/pull/50/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

Comment (stevehadd) [inline model_pipeline/prd_mlops_azml_cluster_hyperdrive_demo.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #4.                           primary_metric_name='mean_absolute_error', 
</pre><p>Are we going to consider other metrics to use as a primary goal for hyperparameter tuning? If so it might be worth creating an issue to remind ourselves of this and record any thoughts we've had so far on doing this (for example some ideas from our discussion at the catchup on 16 June).</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/informatics-lab/precip_rediagnosis/pull/50/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

Comment (stevehadd) [inline model_pipeline/prd_mlops_azml_cluster_hyperdrive_demo.ipynb:1]: <p>What other plots might we want to include here once we have a better dataset? I think this falls under the general theme of what you want to work on in July-Sep 2022, namely thinking about how we evaluate the model performance. Part of that will be what plots do we want, it might be worth adding some next steps to the notebook as part of the planning/thinking about we want to do around this topic,a dn how it would fit into this notebook?</p><p> </p><p>It would be interesting to include some results for different hyperparameters, if one can get the models trained with some different hyperparameters. This might be something we do in subsequent development, rather than in this PR.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/informatics-lab/precip_rediagnosis/pull/50/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='45'/>

Comment (stevehadd) [inline model_pipeline/prd_mlops_azml_cluster_hyperdrive_demo.ipynb:1]: <p>it would be nice to display these result in graphical form perhaps?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/informatics-lab/precip_rediagnosis/pull/50/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='33'/>

**Comment by stevehadd** [comment]
I can also confirm I have successfully rerun the hyperdrive notebook, so all looks to be working OK.

File: model_pipeline/prd_mlops_azml_cluster_hyperdrive_demo.ipynb (hunk 2 above)
Comment (stevehadd) [inline model_pipeline/prd_mlops_azml_cluster_hyperdrive_demo.ipynb:1]: <p>it would be good to a bit more introductory and explanatory test to the notebook, since we're using them to communicate the work we're doing.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/informatics-lab/precip_rediagnosis/pull/50/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (stevehadd) [inline model_pipeline/prd_mlops_azml_cluster_hyperdrive_demo.ipynb:1]: <p>I get an error here, I think this should be updated to X_val</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/informatics-lab/precip_rediagnosis/pull/50/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='43'/>

Comment (stevehadd) [inline model_pipeline/prd_mlops_azml_cluster_hyperdrive_demo.ipynb:1]: <p>when I updated the previous cell to X_val, the resulting plot was just a flat line, so I think something might be wrong here.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/informatics-lab/precip_rediagnosis/pull/50/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='44'/>

File: model_pipeline/prd_cluster_train_demo.py [hunk 3]
```diff
@@ -44,13 +47,20 @@ def main():
 
 
     input_data = prd_pipeline.load_data(prd_ws, args.dataset_name)
-    data_splits, data_dims = prd_pipeline.preprocess_data(input_data, feature_dict)
+    data_splits, data_dims = prd_pipeline.preprocess_data(
+        input_data, feature_dict, 
+        test_fraction=0.2, test_savefn='tmp.csv')  # TODO: update filename! Currently saves local copy of test dataset - perhaps not optimal
```
[the remark this answers is not in the document]
Comment (hannahbrown7) [reply model_pipeline/prd_cluster_train_demo.py:52]: My in line comment isn't every clear, sorry, but it was my intention that the filename would be passed as an argument when calling preprocess_data in the notebook. My point here was more on updating the filename to a filepath with where we actually want to save it and calling the file something more meaningful than tmp.csv. Saving to azure blob sounds like a good idea.

File: model_pipeline/prd_pipeline.py [hunk 4]
```diff
@@ -63,94 +65,150 @@ def build_model(nprof_features, nheights, nsinglvl_features):
...
+        standardScaler.fit_transform(X_train), 
+        columns=X_train.columns,
+        index=X_train.index)
+
+    X_val_scaled = pd.DataFrame(
+        standardScaler.transform(X_val), 
+        columns=X_val.columns,
+        index=X_val.index)
+
+    with open('standardScaler.pkl', 'wb') as fout:
```
[the remark this answers is not in the document]
Comment (hannahbrown7) [reply model_pipeline/prd_pipeline.py:186]: Yes, good point

File: model_pipeline/prd_mlops_azml_cluster_hyperdrive_demo.ipynb (hunk 2 above)
Comment (hannahbrown7) [reply model_pipeline/prd_mlops_azml_cluster_hyperdrive_demo.ipynb:1]: Yes, mainly as simplest. I would like to experiment with the different options though, in particular BayesianParameterSampling and would be good to give GridParameterSampling a go

Comment (hannahbrown7) [reply model_pipeline/prd_mlops_azml_cluster_hyperdrive_demo.ipynb:1]: <p>Yes, I think it would be good to consider other metrics. This to some extent depends on how that model changes and what metrics are sensible for assessing model performance. My understanding is that the primary metric needs to be either the loss function or a metric defined when compiling the model in Tensorflow</p>

Comment (hannahbrown7) [reply model_pipeline/prd_mlops_azml_cluster_hyperdrive_demo.ipynb:1]: I'm not sure if you've been on the experiment(/jobs?) tab on AzureML and looked at one of the HyperDriveRuns but this includes a plot of how different runs with different hyperparameters compare, but maybe we want to visualise this ourselves in a different way 
image

File: model_pipeline/prd_pipeline.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hannahbrown7) [reply model_pipeline/prd_pipeline.py:186]: This is easy to do using mlflow.log_attribute, it doesn't seem quite as straight forward to me to see how to do this with hyperdrive/when running on a cluster, but I could be missing something

File: model_pipeline/prd_pipeline.py [hunk 5]
```diff
@@ -63,94 +65,150 @@ def build_model(nprof_features, nheights, nsinglvl_features):
...
+            df.shape[0], 
+            data_dims_dict['nprof_features'], 
+            data_dims_dict['nheights']
+            ),
+        (0, 2, 1))
+    return df
+
+
+def load_test_data(test_fn, feature_dict, data_dims_dict):
+    """TO DO: where is the best place to store test data and scaler and
```
[the remark this answers is not in the document]
Comment (hannahbrown7) [reply model_pipeline/prd_pipeline.py]: As discussed, this is something to address once update data is available. Create a test dataset saved to blob storage and another train/validate dataset (also saved to blob storage). The train/val data is then what we load each time - then only need to sample train and validate.
